lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,402 @@
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing lapse rate calculation plugins."""
+
+import iris
+import numpy as np
+from numpy.linalg import lstsq
+from improver.utilities.cube_manipulation import enforce_float32_precision
+from improver.constants import DALR
+
+from scipy.ndimage import generic_filter
+
+
+class GenericFunc(object):
+    """
+    The "generic_filter" function extracts the neighbourhood around
+    each point as "buffer". This saves the buffer into "allbuffers"
+    array for the next processing step.
+    """
+    def __init__(self, allbuffers):
+        """Initialise the class.
+
+           Create the global variables that allows the "filter" function
+           to save each extracted buffer into "allbuffers".
+        """
+        # Initialises the iterator:
+        self.i = 0
+        # Saves all the buffers into this array.
+        self.allbuffers = allbuffers
+
+    def filter(self, buffer):
+        """
+        Saves the contents of the buffer into "allbuffers" array so
+        a return value isn't required. However "generic_filter"
+        requires a return value - so use zero.
+
+        Args:
+            buffer (array):
+                Array containing neighourbood points.
+
+        Returns:
+            zero (float)
+                Blank return value required by "generic_filter".
+        """
+        self.allbuffers[self.i, :] = buffer
+        self.i += 1
+        return 0.0
+
+
+class LapseRate(object):
+    """Plugin to calculate the lapse rate from orography and temperature
+       cubes.
+
+    Science background:
+
+    References:
+        The method applied here is based on the method used in the 2010 paper:
+        https://rmets.onlinelibrary.wiley.com/doi/abs/10.1002/met.177
+
+    The current lapse rate (LR) FORTRAN code is kept here:
+    http://fcm9/projects/PostProc/browser/PostProc/trunk/
+                                         model_processing/Fss_CalcTLev1Grad.f90
```

Comment:
Ideally we wouldn't have internal file or intranet paths in publicly sharable code.  Could you remove this please.